server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java

```diff
@@ -789,6 +812,22 @@ private boolean addFailureIfRequiresAliasAndAliasIsMissing(DocWriteRequest<?> re
             return false;
         }
 
+        private boolean addFailureIfRequiresDataStreamAndNoParentDataStream(DocWriteRequest<?> request, int idx, final Metadata metadata) {
+            if (request.isRequireDataStream() && (metadata.hasParentDataStream(request.index()) == false)) {
+                Exception exception = new IndexNotFoundException(
+                    "["
+                        + DocWriteRequest.REQUIRE_DATA_STREAM
+                        + "] request flag is [true] and ["
+                        + request.index()
+                        + "] has no parent data stream",
```

Comment:
I think this message could be a little clearer, a user won't know what a parent data stream is. Perhaps mention something about how the `require_data_stream` parameter requires that an index not be a standalone index?